e2e/playwright/zoo-test.ts

```diff
@@ -0,0 +1,305 @@
+import {
+  test as playwrightTestFn,
+  TestInfo as TestInfoPlaywright,
+  BrowserContext as BrowserContextPlaywright,
+  Page as PagePlaywright,
+  TestDetails as TestDetailsPlaywright,
+  PlaywrightTestArgs,
+  PlaywrightTestOptions,
+  PlaywrightWorkerArgs,
+  PlaywrightWorkerOptions,
+  ElectronApplication,
+} from '@playwright/test'
+
+import {
+  fixtures,
+  Fixtures,
+  AuthenticatedTronApp,
+  AuthenticatedApp,
+} from './fixtures/fixtureSetup'
+
+import { SaveSettingsPayload } from 'lib/settings/settingsTypes'
+export { expect } from '@playwright/test'
+
+declare module '@playwright/test' {
+  interface TestInfo {
+    tronApp?: AuthenticatedTronApp
+  }
+  interface BrowserContext {
+    folderSetupFn: (
+      cb: (dir: string) => Promise<void>
+    ) => Promise<{ dir: string }>
+  }
+  interface Page {
+    dir: string
+    TEST_SETTINGS_FILE_KEY?: string
+    setBodyDimensions: (dims: {
+      width: number
+      height: number
+    }) => Promise<void>
+  }
+}
+
+export type TestInfo = TestInfoPlaywright
+export type BrowserContext = BrowserContextPlaywright
+export type Page = PagePlaywright
+export type TestDetails = TestDetailsPlaywright & {
+  cleanProjectDir?: boolean
+  appSettings?: Partial<SaveSettingsPayload>
+}
+
+// Our custom decorated Zoo test object. Makes it easier to add fixtures, and
+// switch between web and electron if needed.
+const pwTestFnWithFixtures = playwrightTestFn.extend<Fixtures>(fixtures)
+
+// In JavaScript you cannot replace a function's body only (despite functions
+// are themselves objects, which you'd expect a body property or something...)
+// So we must redefine the function and then re-attach properties.
+type PWFunction = (
+  args: PlaywrightTestArgs &
+    Fixtures &
+    PlaywrightWorkerArgs &
+    PlaywrightTestOptions &
+    PlaywrightWorkerOptions & {
+      electronApp?: ElectronApplication
+    },
+  testInfo: TestInfo
+) => void | Promise<void>
+
+// The below error is due to the extreme type spaghetti going on. playwright/
+// types/test.d.ts does not export 2 functions (below is one of them) but tsc
+// is trying to use a interface name it can't see.
+// e2e/playwright/zoo-test.ts:64:14 - error TS4023: Exported variable 'test' has
+// or is using name 'TestFunction' from external module
+// "/home/lee/Code/Zoo/modeling-app/dirty2/node_modules/playwright/types/test"
+// but cannot be named.
+export const test = (
```

Comment:
For the type errors here, I confirmed that `export`ing SuiteFunction and TestFunction fix the errors. However, in Playwright `main`, these types have already been deleted in https://github.com/microsoft/playwright/pull/33142, so I don't think it's worth doing gymnastics to try to make this work. Can we use `any` and move on? Perhaps when we upgrade Playwright in the future, we can revisit this.

Reply:
I found that upgrading Playwright to 1.48.2 or later resolved the type error! 😃 So maybe these comments can actually get deleted.

Reply:
Whoops yep will delete - thanks for checking that out man!